internal/model/transform.go

```diff
@@ -326,6 +339,113 @@ func (m *Model) instantiateStruct(
 	return newType, nil
 }
 
+func (m *Model) instantiateChoice(
+	pkgScope *ast.Package,
+	templChoice *ast.Choice,
+	templateArguments []*ast.TypeReference,
+	instantiatedName string) (*ast.TypeReference, error) {
+	if len(templateArguments) == 0 {
+		return nil, errors.New("choice instantiation called without parameters")
+	}
+	if len(templChoice.TemplateParameters) != len(templateArguments) {
+		return nil, ErrIncorrectNumberOfTemplateParameters
+	}
+	// templateTypes is a map from the template parameter name to the instantiated
+	// type.
+	templateTypes := map[string]*ast.TypeReference{}
+	for ix, ta := range templateArguments {
+		// Instantiate the subtype
+		ta, err := m.instantiateType(pkgScope, ta, ta.TemplateArguments, "")
+		if err != nil {
+			return nil, err
+		}
+		templateTypes[templChoice.TemplateParameters[ix]] = ta
+	}
+
+	newType := &ast.TypeReference{
+		IsBuiltin: false,
+		Package:   pkgScope.Name,
+		Name:      instantiatedName,
+	}
+	if _, exists := pkgScope.Choices[newType.Name]; !exists {
```

Comment:
It seems that we can invert an if here. The code inside the statement is a bit long and it becomes hard to read this way.